### ISSUE TITLE
Milestones for renaming g0-merge to main

### ISSUE THREAD

**Issue by ammarhakim** [ammarhakim/gkyl#134]
We are nearly ready for renaming g0-merge to main and using that for all simulations. However, we are not yet there yet. Here is a list of milestone we should meet before we do this rename. Please comment/discuss.

- Remove all unused C++ kernels and code from the system
- Cleanup usage of ZeroArray so we do not need to check for GPUs every time we make a call to the G0 code. This is very confusing and I do not think it is really needed.
- Cleanup and simplify CartField. Many features are not needed and so remove them.
- Switch to Git version of LuaJIT so we are on the latest LJ code and not some older, forked version.
- Remove  dependencies that are no longer needed for G2. For example, is Eigen needed? I not, lets remove. (The dispersion Tool use Eigen but I think this is easy to replace  with OpenBLAS functions).
- Re-enable automatic regression tests to ensure we have not broken anything. Even if this is on local machines we need to do this before we rename.

A secondary thing we can do but perhaps can wait is:

- Wrap Moments code in using Lua-C API (no FFI but straight C). We will need this for fluid sims very soon.

**Comment by manauref** [comment]
@ammarhakim some responses to your proposals:
1. We might want to keep **some** C++ kernels in g2, for two reasons: a) there's no pressure to move them to g0 now (e.g. HasegawaWakatani), b) To leave a trace of of how g2 C++ code can be added (for dev perhaps).
2. Clean up usage of ZeroArray: sure, easy, I think we only check for GPUs in a couple of places.
3. Clean up CartField: I've done a lot of that already. The vast majority of what's there we still use, but there may be a couple things we can remove still.
4. LuaJIT: you are better suited for that.
5. Dependencies: Aside from Eigen, I don't know what else to remove. Eigen: only one place in twist-shift uses it, and we can easily remove that.
6. Reg tests are running. Not all of course, but the key ones (ESGK and Vlasov).

**Comment by manauref** [comment]
I'd like to add 2 to-do's:
1. Remove the dependency on MG Poisson solver code. There's an ES energy function in there that we are using in VP and GK.
2. Some volume integrals (CartFieldIntegratedQuantCalc and calculation of ES energy) still have MPI inside of them. Abstract that out to Messenger instead.

**Comment by ammarhakim** [comment]
As we discuss the merge, I would like to add the following two milestones. Once this initial discussion is complete we will make a proper list so we know what needs to be done for the merge.

- Add a install script to mkdeps in G2 to build GkeyllZero.
- Get the regression system running again and look at the regression tests (work should be split between people that can properly certify sims). We should look into hooking the regression test on some systems to run semi-automatically.

**Comment by manauref** [comment]
We can do 1, but 2 is not needed for the merge. No one is checking
regression tests in g2/master, so that is no better than what is happening
in g2/g0-merge and is likely worse, because we've been working with
g0-merge reg tests while master's have been neglected.

**Comment by manauref** [comment]
I'm saying merging is not a regress in that regard. If anything it is an
improvement, even if it doesn't (yet) have the testing that you want.
